Import OPML from URL

yaelirub opened the pull request:
Fixes #442 

Added the option to import podcasts from an opml file URL. 

https://user-images.githubusercontent.com/1335657/220860852-f56eab0b-fbab-4f9e-8840-0ec06d12f2f8.mov


A couple of notes 
- The icon for the "Import from URL" is temporary as I'm waiting on design. I will update it and the release notes once available.
- `opmlImporter` class logic for importing the podcasts from the opml file is strongly coupled with a progress view which is not ideal, especially now since we're calling it from SwiftUI view. That's why I had to make the progress view optional and change a few more things there. It would be good for us to refactor that at some point in the future. 

## To test
**from Profile**
1. Go to settings 
2. Go to import podcasts 
3. Tap import from URL
4. tap an invalid url 
5. see error "OPML import failed"
6. enter a valid url, but not opml 
7. see error "OPML import failed"
8. enter valid opml url
9. see "OPML import succeeded"
10. See button title change to "Done"
11. tap "Done"
12. see view dismissed 

**from Onboarding**
1. signup 
2. Tap Import podcasts from welcome screen 
3. repeat steps 1-12 from above section

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

yaelirub committed:
Added a row in import screen to allow importing an OPML from URL.
- Icon for Import by URL is temporary and will be replaced by icon from design

yaelirub committed:
Added description for import from URL screen

yaelirub committed:
Added textfield and a button to OPML from URL screen

yaelirub committed:
Updated Loading alert of OpmlImport to optional.
OpalImport is strongly coupled with LoadingAlert, we don't always want to show that alert.

yaelirub committed:
Added logic to import from url
- we already have an existing function to import podcasts from, so once we download the opal to a local file, I'm using the existing function to turn the data into podcasts and add to the podcasts list.

yaelirub committed:
Posting a notification on opal import failure from OpmlImport

yaelirub committed:
Added a visual loader in Import DetailsView
- Changed button to look like other screens
- updated text and added new localized strings

yaelirub committed:
decluttering code

yaelirub committed:
update button title and action on import success
- button will now will have title "Done" and will dismiss the view on tap

yaelirub committed:
fix - posting the right notification

yaelirub commented:
I probably need to update analytics

yaelirub committed:
Updated icon and added release note

yaelirub commented:
Added the new icon and release notes.

leandroalonso commented on podcasts/Onboarding/Import/ImportDetailsView.swift:
```diff
@@ -1,7 +1,17 @@
+        case none
+        case success
+        case failure
+    }
```
[nit] Wdyt about add breaklines to keep the `enum` separated from the `@Environment` and then another breakline separating the `@State`?

yaelirub replied:
updated

leandroalonso commented on podcasts/Onboarding/Import/ImportDetailsView.swift:
```diff
@@ -24,33 +34,99 @@ struct ImportDetailsView: View {
+            case .none: Text("")
+            case .success:
+                Text(L10n.opmlImportSucceededTitle)
+                    .foregroundColor(Color.green)
```
We should use a dynamic color here to avoid weirdness if the user is using any different theme, I believe `support02` would be the correct one here.

yaelirub replied:
updated

leandroalonso commented on podcasts/Onboarding/Import/ImportDetailsView.swift:
```diff
@@ -24,33 +34,99 @@ struct ImportDetailsView: View {
+                    .foregroundColor(Color.green)
+            case .failure:
+                Text(L10n.opmlImportFailedTitle)
+                    .foregroundColor(Color.red)
```
Same as above, here we should probably use `support05`

yaelirub replied:
updated

leandroalonso commented on podcasts/Onboarding/Import/ImportDetailsView.swift:
```diff
@@ -1,7 +1,17 @@
 import SwiftUI
 
 struct ImportDetailsView: View {
+    enum OPMLImportResult {
```
Wdyt about renaming that to `OPMLURLImportResult`? Just so we make sure to convey that this is for external URLs.

yaelirub replied:
updated

leandroalonso commented on podcasts/Onboarding/Import/ImportViewModel.swift:
```diff
@@ -34,10 +34,11 @@ class ImportViewModel: OnboardingModel {
+        .init(id: .opmlFromURL, displayName: "URL", steps: L10n.importOpmlFromUrl)
     ]
 
     enum ImportAppId: String, AnalyticsDescribable {
```
Now this is not only related to apps, maybe renaming it to `ImportSourceId` makes more sense? Wdyt?

yaelirub replied:
That lead me to replace all references to app to import source. Instead of installedApps, Im looking for AvailableResources

leandroalonso reviewed (commented):
@yaelirub works as described, and left a few suggestions to improve the code.

Also, a question: when an OPML is successfully imported, we show a success message and a button. When I tap the button, I return to the previous screen (Take your podcasts with you...)

Wdyt about dismissing the modal as a whole?

yaelirub commented:
> when an OPML is successfully imported, we show a success message and a button. When I tap the button, I return to the previous screen (Take your podcasts with you...)

@leandroalonso , the reason I didn't want to dismiss the whole modal is because when coming from onboarding, I wanted them to go back to the welcome screen, in case they wanted to do something else there. Do you think its still worth coming back to the welcome screen from onboarding, or should we dismiss the whole modal? 

Actually, I'll just dismiss the full modal and see if you think it makes sense for both flows

yaelirub committed:
Update according to code review
- changed all app references to importSource as we have more than apps now. That lead me to remove the special treatment  for handling no app installed as we're always going to have .others and import from url

yaelirub commented on podcasts/Onboarding/Import/ImportLandingView.swift:
```diff
@@ -5,11 +5,6 @@ struct ImportLandingView: View {
     let viewModel: ImportViewModel
 
     var body: some View {
-        // It's possible the user has no supported apps installed
```
No need to check that anymore. We will always have at least 2 available resources: others and opml from url

yaelirub commented:
@leandroalonso , ready for another look

leandroalonso reviewed (approved):
@yaelirub nice job! :shipit:

emilylaguna commented on podcasts/Onboarding/Import/ImportDetailsView.swift:
```diff
@@ -1,56 +1,134 @@
+        VStack {
+            TextField("https://...", text: $opmlURLText)
+                .textFieldStyle(RoundedBorderTextFieldStyle())
+                .autocapitalization(.none)
```
I noticed the field isn't styled with theme and can have reduced readability depending on the system dark/light setting.

If we update to use the [`requiredStyle`](https://github.com/Automattic/pocket-casts-ios/blob/aee2844a9ed35524b897f75198999fd0cf9bf0be/podcasts/Themeable%2BSwiftUI.swift#L26) from Themeable+SwiftUI.swift we can apply the standard theme to the textfield.

<details>
  <summary>We also get highlighting the textfield in red if there's an error for free. </summary>
  
<img width="200" src="https://user-images.githubusercontent.com/793774/221645530-071e26cb-07f3-4f89-857a-ffef29808081.png">
</details>

Wdyt?

| Before | After |
|:---:|:---:|
|<img width="200"  src="https://user-images.githubusercontent.com/793774/221645206-049b9500-f71b-4ae6-9ee4-814c5dae2f57.png">|<img width="200" src="https://user-images.githubusercontent.com/793774/221645275-a089134e-3df7-43c3-9e64-1e5c2543af5b.png">|

```suggestion
            TextField("https://...", text: $opmlURLText)
                .requiredStyle(OPMLURLImportResult == .failure)
                .autocapitalization(.none)
```

emilylaguna replied:
Wdyt about defaulting to the URL keyboard type?

```suggestion
                .autocapitalization(.none)
                .keyboardType(.URL)
```

emilylaguna commented on podcasts/Onboarding/Import/ImportViewModel.swift:
```diff
@@ -142,9 +154,40 @@ extension ImportViewModel {
+            }
+
+            let documentsDirectory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
+            let fileURL = documentsDirectory.appendingPathComponent("feed.opml")
```
[nit] I think it makes more sense to use a temporary directory here instead of the documents since we only need access to the file during the import process and without a cleanup step after the import, the feed.opml will continue to exist on the users device indefinitely. 

Using the temporary directory will let the system remove it whenever it needs to and we don't need to worry about cleaning it up. 

```suggestion
            let temporaryDirectory = FileManager.default.temporaryDirectory
            let fileURL = temporaryDirectory.appendingPathComponent("feed.opml")
```

emilylaguna reviewed (changes requested):
Excellent work on this, sorry I'm reviewing a bit later. I gave it a test and it imported like a charm!

I left a couple small comments, and one item about updating the text field to be consistent with the theme.

yaelirub committed:
Update podcasts/Onboarding/Import/ImportViewModel.swift

Co-authored-by: Emily Laguna <793774+emilylaguna@users.noreply.github.com>

yaelirub committed:
Updated according to code review

yaelirub committed:
Merge branch 'feature/442_import_opml_from_url' of github.com:Automattic/pocket-casts-ios into feature/442_import_opml_from_url

yaelirub commented:
Thank you @emilylaguna , updated and ready for another look

emilylaguna commented on podcasts/Onboarding/Import/ImportDetailsView.swift:
```diff
@@ -1,56 +1,135 @@
+
+    private var opmlViewButton: some View {
+        Button(action: {
+            if opmlURLImportResult == .success {
```
I think we can add a return here to prevent the logic from falling through and kicking off the download again.

emilylaguna reviewed (approved):
Awesome work. 🥳 

I noticed one last thing while giving this another test, after I've finished my import and I tap the Done button the download kicks off again while the view is being dismissed and I get another log for 'File downloaded to...'

Approving now since after that's fixed we'll be good to merge :shipit:

yaelirub committed:
added according to code review
